extra/mediawiki/api/api.factor

```diff
@@ -0,0 +1,216 @@
+        ] }
+    } cond ] assoc-map ;
+
+: <api-request> ( params -- request )
+        {
+            { "format" "json" }
+            { "formatversion" 2 }
+            { "maxlag" 5 }
+        } swap assoc-union prepare
+        endpoint get
+    <post-request>
+        contact get vm-version vm-git-id 7 head
+        "%s Factor/%s %s mediawiki.api" sprintf "User-Agent"
+        set-header ;
+
+: oauth-post ( params -- response data )
+    oauth-login get
+        dup consumer-token>>
+        over consumer-secret>> <token> consumer-token set
+        dup access-token>>
+        swap access-secret>> <token> access-token set
+    <api-request>
+        <oauth-request-params> set-oauth
+    http-request ;
+
+: cookie-post* ( params -- assoc )
+    <api-request>
+        cookies get >>cookies
+    http-request [ cookies>> cookies set-global ] dip json> ;
+
+: login-token ( -- token )
+    {
+        { "action" "query" }
+        { "meta" "tokens" }
+        { "type" "login" }
+    } cookie-post*
+    "query" "tokens" "logintoken" [ of ] tri@ ;
+
+: login ( -- cookies )
+    [
+        "login" "action" ,,
+        password-login get dup username>> "lgname" ,,
+        password>> "lgpassword" ,,
+        login-token "lgtoken" ,,
+    ] { } make cookie-post* drop cookies get ;
+
+: cookie-post ( params -- response data )
+    <api-request>
+        cookies get [ login ] unless* >>cookies
+    http-request ;
+
+: anon-post ( params -- response data )
+    <api-request> http-request ;
+
+: code-200? ( response assoc -- ? )
+    over code>> dup 200 = dup [ 3nip ] [
+        -roll "http status code %d" printf
+        swap header>> [ "=" glue print ] assoc-each
+        ...
+        10 minutes sleep
+    ] if ;
+
+: retry-after? ( response -- ? )
+    header>> "retry-after" of dup [ dup seconds sleep ] when ;
+
+: nonce-already-used? ( assoc -- ? )
+    "error" of
+    [ "code" of "mwoauth-invalid-authorization" = ]
+    [ "info" of "Nonce already used" swap subseq-start ] bi
+    and ;
+
+: readonly? ( assoc -- ? )
+    "error" "code" [ of ] bi@ "readonly" = dup
+    [ 5 minutes sleep ] when ;
+
+: failed? ( response assoc -- response assoc ? )
+    2dup 2dup code-200? not
+    rot retry-after? or
+    over nonce-already-used? or
+    swap readonly? or ;
+
+: dispatch-call ( params -- response data )
+    {
+        { [ oauth-login get ] [ oauth-post ] }
+        { [ password-login get ] [ cookie-post ] }
+        [ anon-post ]
+    } cond ;
+
+PRIVATE>
+
+: api-call ( params -- assoc )
+    f f [
+        failed?
+    ] [
+        2drop dup dispatch-call
+        [ json> ] [ swap print rethrow ] recover
+        "warnings" "errors" [ over at [ ... ] when* ] bi@
+    ] do while 2nip ;
+
+<PRIVATE
+
+:: (query) ( params -- obj assoc )
+    { { "action" "query" } } params assoc-union api-call dup
+    dup "query" of [ nip ] when*
+    "siprop" params key? [
+        params "prop" "list" "meta" [ of ] tri-curry@ tri or or
```

Comment:
``params { "prop" "list" "meta" } [ of ] map-find drop``

Maybe?

Reply:
```
params "prop" "list" "meta" [ of ] tri-curry@ tri or or
params { "prop" "list" "meta" } [ of ] with map-find drop
params { "prop" "list" "meta" } values-of sift first
params { "prop" "list" "meta" } intersect-keys values first
```
I think I will use the third one